setup.sh

```diff
@@ -0,0 +1,74 @@
+#! /usr/bin/env bash
+
+# Bash strict mode
+set -euo pipefail
+
+echo "PORT=8083" > .env
+echo "EXTRA_PORT='8082'" >> .env
+echo "WEBPACK_DEV_URL=http://0.0.0.0:8082" >> .env
+
+echo "##########################"
+echo "#                        #"
+echo "#         Github         #"
+echo "#                        #"
+echo "##########################"
+echo "You need to register a new OAuth application:"
+echo "https://github.com/settings/applications/new"
+echo
+echo "You can put those parameters:"
+echo "* Application name: BSI development"
+echo "* Homepage URL: 0.0.0.0:8083"
+echo "* Application description: This a development OAuth application for BSI"
+echo "* Authorization callback URL: http://0.0.0.0:8083/auth/verify"
```

Comment:
Yeah it should probably be 127.0.0.1. Although 0.0.0.0 actually works fine in every browser I've ever seen.

Reply:
I'm pretty sure that neither will work if you happen to be running the webapp in a container but the browser in your normal system, for instance; which is among the reasons I think it's a good idea to ask.

Reply:
I haven't had problems with using 0.0.0.0 before, I am interested what the specific worries would be here. We run services in containers with 0.0.0.0 in many places and it works well but I could very possibly be ignorant to the problem.

Reply:
Listening on 0.0.0.0 (i.e. any IPv4 address) is fine, but this isn't a listen address, it's an instruction to arrange for the browser to be redirected to that URL, and the browser won't know where to connect to if it's sent to 0.0.0.0.

Reply:
I don't understand that. If you can enter http://0.0.0.0 in a browser to load a web page, why can't a service also ask the browser to redirect to that URL?

I do think we should probably be more specific and say 127.0.0.1 - but Colin did you say you also don't like that idea?

Reply:
Listening on 0.0.0.0 means that the application is prepared to accept connections on any address.  It doesn't mean that the browser will have the slightest clue how to send packets to the application if you just say `http://0.0.0.0`.  (I have just tested this, running BSI in a container; while it says `http://0.0.0.0:8000` on startup, as expected, when I copy that into my browser on the system hosting the container in question, that URL does not work.  Based on an elementary understanding of TCP/IP, there is really no reason to expect that it *should* work.)

`127.0.0.1` is also no good, because if you're running the system in a container then `127.0.0.1` does not address the container!  You need to ask.

Reply:
Let me clarify a bit further: `http://0.0.0.0:8000/` or similar will probably work if your application happens to be running in the same OS context as your browser.  But if you're running the application in a container and your browser in your host OS - a perfectly sensible configuration, especially if you work on lots of applications and don't want to clutter your host OS with all their possibly-incompatible dependencies, or don't want to be stuck running the same OS version as is required by all these applications - then it will not work.

Reply:
When you tell an app to listen on `0.0.0.0` it means "listen for all connections". But when you ask a browser to find `0.0.0.0` it interprets it as localhost (127.0.0.1) because, as you say, what else could it be?

When we run our docker containers for local development, we expose, or publish, the container's port on a host port. So if the application is running on, say, port 8000 inside the container, we will simply publish it to port 8000 on the host. This is why browsing (or redirecting) to http://0.0.0.0:8000 will work without issue.

Reply:
That's one possible setup, sure.  But if you're requiring that, it's currently entirely undocumented in this project that you're doing so ...

Reply:
(Also, given the popularity of port 8000, this is a somewhat problematic requirement if you're working on more than one application at once.)

Reply:
@cjwatson Yes we're not actually going to use port 8000. And this script is intended as a helper script for our recommended local development flow, for which we will know how we're running the script. If you don't want to use that flow, you're welcome to set your own environment variables and run the project through another means. Hopefully we'll even document how to do that as well.

If you'd like to talk to me about [how we run projects in general](https://github.com/canonical-webteam/practices/blob/master/local-development/the-run-script.md), we can do that separately, but I don't think we should delay this PR being merged in the meantime.

Reply:
Well, a simple comment about the port-forwarding setup would satisfy me in this case, since then a confused developer could look at this script and have a breadcrumb trail to follow.  Why not do that?

Reply:
Sounds good. @tbmb could we do that?

Reply:
I agree with that but this should be done in the run script and not in the project...
Because here the run script those this : 
```
        ${run_function} "--env PORT=${PORT} --publish ${PORT}:${PORT} ${publish_extra_ports} ${detach} ${run_serve_docker_opts} ${run_options}" ${run_command} $*
```

So it publishes the PORT to PORT, this would need a big modification to say that every project should run with the port 8000 and then forward the port in the run script...